kotlintest-runner/kotlintest-runner-jvm/src/main/kotlin/io/kotlintest/runner/jvm/spec/SpecExecutor.kt

```diff
@@ -0,0 +1,87 @@
+package io.kotlintest.runner.jvm.spec
+
+import arrow.core.Try
+import io.kotlintest.IsolationMode
+import io.kotlintest.Project
+import io.kotlintest.Spec
+import io.kotlintest.runner.jvm.TestEngineListener
+import io.kotlintest.runner.jvm.internal.topLevelTests
+import org.slf4j.LoggerFactory
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+
+/**
+ * Handles the execution of a single [Spec] class.
+ * Delegates to a [SpecRunner] which determines how and when
+ * to instantiate fresh specs based on the [IsolationMode] of the spec.
+ */
+class SpecExecutor(private val engineListener: TestEngineListener,
+                   private val listenerExecutors: ConcurrentLinkedQueue<ExecutorService>,
+                   private val scheduler: ScheduledExecutorService) {
+
+  private val logger = LoggerFactory.getLogger(this.javaClass)
+
+  private fun withListenerExecutor(thunk: (ExecutorService) -> Unit) {
+    val listenerExecutor = listenerExecutors.poll() ?: Executors.newSingleThreadExecutor()
+    thunk(listenerExecutor)
+    listenerExecutors.add(listenerExecutor)
```

Comment:
Should this be before thunk?
I'm thinking that if `thunk` fails, we'll have a dangling executor that might not be garbage collected if it's executing something